### ISSUE TITLE
Extends support

### ISSUE THREAD

**Issue by pi0** [nuxt/framework#3222]
Feature:
- [x] Initial support (#3008)
- [x] `components/` dir (#3108)
- [ ] `composables/` dir
- [ ] `plugins/` dir
- [ ] `pages/` dir
- [ ] `middleware/` dir
- [ ] `server/` dir
- [ ] Merge `hook` config key

Documentation:
- [ ] Basic usage 
- [ ] Theming
- [ ] Multi-app

Upstream: 
- [nitropack] hooks for extending dirs
- [c12] install cloned dependencies

**Comment by Tahul** [comment]
Hey there 👋

We've made a review of `extends` feature with @kevinmarrec so I can document it!

I would like to brainstorm few ideas with you, before beginning any changes.

Aside from these points, we've tried the extends feature together with @kevinmarrec and it went really well!

I will start creating a PR to document this with appropriate examples as you mentioned in the first message. 🙂

### Extends configuration

With the current `extends` implementation, once you pass a path to be extended, Nuxt will try to grab every available feature from this extended directory.

It is really great to have such behavior as it makes `extends` a real plug & play feature, but having some flexibility over how we do extend our project would be really great.

What I suggest is to have a similar way to configure your extends as components does:

```ts
defineNuxtConfig({
   extends: [
		'./base', // Would be fully extended
		{
			path: './ui',
			middleware: false, // Would skip middleware extending
			pages: false, // Would skip pages extending
			// All of the rest are set to `true` by default
		}
   ]
})
```

That would maybe include a PR to `c12` package in order to support this.

### Extends pages prefixing

One cool feature would be to be able to rewrite the base path of pages from an extend target.

Let's say you extend your project from a `blog` theme that has default pages exposed.

The blog theme might expose index pages and such from the root of its `pages` directory, so that would be overwritten by your own website.

What you might want then is to have these blog pages nested under a `/blog` prefix in your project.

We could use the previous configuration object to specify this:

```ts
defineNuxtConfig({
   extends: [
		'./base', // Pages would be merged with project-level pages
		{
			path: './blog',
			pagesPrefix: '/blog' // Pages would be nested under `/blog` route prefix
		}
   ]
})
```

### Low to high instead of high to low priority management

One thing we noticed during the review of `extends` is that currently the priority is handled in a `high to low` manner:

```ts
defineNuxtConfig({
   extends: [
		'./ui', // This will override `./base`
		'./base'
   ]
})
```

This is nitpicking but I think this is an opposite pattern of what we commonly expect from extends features.

For instance [`tsconfig.json`](https://www.typescriptlang.org/tsconfig#extends) or [`.eslintrc`](https://eslint.org/docs/user-guide/configuring/configuration-files) files has this "last has highest precedence" as convention.

WDYT @danielroe @pi0 @Atinux ? 😄

**Comment by Atinux** [comment]
- Extends options

What about using the array syntax for it since the first argument is required?

```ts
defineNuxtConfig({
   extends: [
     './base', // Pages would be merged with project-level pages
     ['./blog', { pages: { prefix: '/blog' } }],
   ]
})
```

- Low to hight makes sense actually, I cannot recall why we have the opposite order (Pooya may be able to answer)

**Comment by Atinux** [comment]
What about using `from` key:

```ts
defineNuxtConfig({
   extends: [
     './base', // Pages would be merged with project-level pages
     {
        from: './blog',
        pages: { prefix: '/blog' },
        middleware: false
     }
   ]
})
```

**Comment by kevinmarrec** [comment]
- Last proposal for **extends options** from @Atinux sounds good to me (using `from` key)
- I also agree for **Low to High** order, but it needs to be consistent within `nitropack` & `c12` first as stated by @pi0 here : https://github.com/nuxt/framework/pull/3783#discussion_r832482516

For both above points, changes need to be applied to `c12`, so I'll let @pi0 handles that.
I will be able to take the lead on implementing **extends options** whenever `c12` can ship the options of `extends` objects in the extended `layers` (I can think of adding the options here: https://github.com/unjs/c12/blob/main/src/loader.ts#L11 under a specific key).

**Comment by pi0** [comment]
Thanks for the feedback @Tahul @Atinux @kevinmarrec 💚 

Overriding by layer seems a good idea. I like both `[from: string, override: T]` and `{ from: string, override: T }` syntaxes. Note that we cannot reuse the whole object syntax namespace for overriding. c12 might need it for additional source options in the future.  (~> https://github.com/unjs/c12/issues/9)

In order to disable scanning of a directory of a layer, I think we can reuse of `dirs` options with `false` value. `dirs: { pages: false }` and by extending each layer or from within layer itself, we can disable it.

Regarding prefixes, I made lots of thoughts since (https://github.com/nuxt/rfcs/issues/30). Extensions will be surely the basis for multi-app support but it is more complicated than you might imagine and too soon implemented in this stage. Usage of baseURL and `<NuxtLink>` in the prefixed layer will be broken when the layer is using path like `/post/1` as `pages/post/[id].vue` and we prefix it with `/blog`. 


Regarding the ordering convention, I guess we will keep it as is. [defu@v6](https://github.com/unjs/defu/releases/tag/v6.0.0) is released to respect the same convention for joining arrays. Internally it makes implementation of multi-layer much more straightforward since we can simpy iterate and in next layers, ignore if a duplicate item exists (vs resolve and override with reverse manner). User wise, `extends` actually means extending by defaults of sublayers and visually, a higher layer overrides the ones below.

**Comment by Tahul** [comment]
Hey, thank you a lot for your feedback @pi0 🙏

I noticed the existence of an additional file/directory inside Nuxt 3 structure: `app/`

I saw there here: https://github.com/nuxt/framework/pull/3485 & [here](https://v3.nuxtjs.org/docs/directory-structure/pages#router-options).

I think as these file/this directory contains data configuring the app, we should allow extending of this file aswell?

I'm using `app/router.options.ts` to extends the `scrollBehavior` from it.

**Comment by pi0** [comment]
Sure thing. Makes sense for supporting `app/` extension.

**Comment by fanckush** [comment]
@pi0 will `/assets` also be supported?

**Comment by pi0** [comment]
> will /assets also be supported?

It fall into named layer aliases. `assets` is basically a standard directory but we are not automatically scanning it. For layers, it would be `~layer/assets` to access assets of each layer.

**Comment by dm-heinze** [comment]
Hey guys :) Really great feature!
Is there a possibility to extend the configs via module?

I already tried this, but it doesn't work. I guess because the config is already loaded at the time my module setup is processed:
```
async setup (options, nuxt) {
    const runtimeDir = fileURLToPath(new URL('./runtime', import.meta.url))
    nuxt.options.extends = runtimeDir
}
```

**Comment by pi0** [comment]
@dm-heinze You can load and extend options in modules by modifying `nuxt.options` as before within the module but you have to take care of normalization as they are already done when we reach modules.

If I correctly get your idea, you want to make it easier for modules by simply following Nuxt convention for `runtime` dir right? That's something we can definitely work on! (Only note that we can extend by directory structure and not an additional `runtime/nuxt.config`.)

**Comment by sawden** [comment]
Can the extend feature be used for different themes that change some parts of the application layout?

**Comment by pi0** [comment]
Can you please explain more @sawden? Two themes can provide or override composables and components of course used in layout.

**Comment by sawden** [comment]
@pi0 
I want to build an app that has its own theme for each customer. The functionality is similar for all of them and there are only small differences such as more filter options or a different way of displaying lists, for example.

I designed a small example to show what I mean:

![example app1](https://user-images.githubusercontent.com/36924392/179032453-d1590033-f7bb-4128-99ac-950c1f3341ef.png)

**Comment by dm-heinze** [comment]
Hey, what about extending components that are dynamically imported? Will the named layer alias work for them as well? 

For example:
```js
const AsyncComp = defineAsyncComponent({
  loader: () => import('~layer/components/Foo.vue')
})
```

**Comment by Tahul** [comment]
@kevinmarrec ; we added a to-do for `app.vue` support :)

Let's follow this issue to track the task!

**Comment by pbender87** [comment]
> > will /assets also be supported?
> 
> It fall into named layer aliases. `assets` is basically a standard directory but we are not automatically scanning it. For layers, it would be `~layer/assets` to access assets of each layer.

Very cool feature. Is there any update? And does this feature also work for preprocessor Options like :
`vite: {
    css: {
      preprocessorOptions: {
        scss: {
          additionalData: `@import "~layer/assets/css/variables.scss";`,
        },
      },
    },
  },`

**Comment by HigherOrderLogic** [comment]
Maybe it's a good idea to add an `additionalFolders` to copy other folders apart from standard directories. Ex:
```ts
defineNuxtConfig({
	extends: [
		{
			from: "github:Atinux/content-wind-template",
			additionalFolders: ["content"],
		}
	]
});
```

**Comment by manniL** [comment]
@SuleymanDeveloper Please open a *new* issue and provide a reproduction ☺️

**Comment by BracketJohn** [comment]
Hey people 👋 

Thanks for pushing on layers - I've experimented around with them for a bit and noticed that middleware-ordering does not seem to be supported for layers: When I create a layer with a middleware `0.test.ts`:
```ts
// file: `0.foo.ts`
export default eventHandler((event) => {
  console.log('foo')
})
```

And then a middleware `1.bar.ts` in my app that `extends` from the layer:
```ts
// file: `1.bar.ts`
export default eventHandler((event) => {
  console.log('bar')
})
```

The log output in the app is:
```ts
bar  // -> 1.bar.ts
foo  // -> 0.foo.ts
```

So 1 comes before 0.

As development is ongoing on layers, maybe this ordering problem can still be taken into account for development. Ordering like this has many use-cases, e.g., when we want to add a `session` to the `event` and make sure that all subsequent middlewares get access to the session. See also https://github.com/nuxt/framework/issues/9396 for an issue on ordering for modules.

I guess one possible work-around at the moment would be to shift around layers using `nuxt.options._layers` (https://nuxt.com/docs/guide/going-further/layers#multi-layer-support-for-nuxt-modules).

**Comment by pi0** [comment]
@BracketJohn Order is respected in each layer (layer1/0.foo > layer1/0.bar) and higher layers always have priority for scanning behavior (layer1/999.foo > layer2/1.bar).

Using `app:resolved` hook to reorder scanned middleware in a more customized manner would be one solution.

**Comment by kvanska** [comment]
Would it be possible to configure merging of data files from base layer? For example:

base layer:
`
// file: /locales/en.json:
{
  "validations":{
    "_default": "{field} is not valid"
  }
}
`

app:
`
// file: /locales/en.json:
{
  "title": "My amazin site"
}
`

merged file:
`
// file: /locales/en.json:
{
  "title": "My amazin site",
  "validations":{
    "_default": "{field} is not valid"
  }
}
`

**Comment by dakt** [comment]
Prefixing would be great since it would give us centralised name-spacing which I find very important on larger projects with many teams.

**Comment by AndrewBogdanovTSS** [comment]
@dakt prefixing is not always desired behavior, right now there are a lot of cases where overwriting is a more desirable behavior and you can easily implement a "prefixed" mode by just creating a folder or naming your files differently

**Comment by dakt** [comment]
@AndrewBogdanovTSS  sure it's not always desired, just like any form of namespaceing isn't always desired. Perhaps on simple projects.
Problem with `you can easily implement a "prefixed" mode by just creating a folder or naming your files differently` is that namespacing is hidden in naming convention rather then in one, centralised space. This makes namespacing really difficult  and prone to errors on large projects.

**Comment by ErosBurastero** [comment]
> 

Somebody knows if it is necessary to provide an open project or i can use a private project to extend it ? 
Because i am trying to extend a private project but it doesn't work, thanks

**Comment by ennioVisco** [comment]
> > will /assets also be supported?
> 
> It fall into named layer aliases. `assets` is basically a standard directory but we are not automatically scanning it. For layers, it would be `~layer/assets` to access assets of each layer.

This would be quite a cool feature, but even simpler than that, layers would be much more usable if relative paths within the layer are correctly parsed as being within the layer, and not relative to the user's path

**Comment by pi0** [comment]
> This would be quite a cool feature, but even simpler than that, layers would be much more usable if relative paths within the layer are correctly parsed as being within the layer, and not relative to the user's path

It is not easily possible to universally do this as each module/feature scanning is different. Howver indeed we can do it at least for common built-in options such as `plugins` and `modules` to preresolve relative bases before merging 👍🏼

**Comment by ennioVisco** [comment]
> > This would be quite a cool feature, but even simpler than that, layers would be much more usable if relative paths within the layer are correctly parsed as being within the layer, and not relative to the user's path
> 
> It is not easily possible to universally do this as each module/feature scanning is different. Howver indeed we can do it at least for common built-in options such as `plugins` and `modules` to preresolve relative bases before merging 👍🏼

I can understand... What about setting a custom alias? Maybe even inside the original layer `nuxt.config.ts`, something like: `layerAlias: @layer1`, to be used also in paths within the layer...

**Comment by ErosBurastero** [comment]
I've tried to extend a private repositorie from github by creating a token and setting it in an enviroment variable as documentation implies, both windows terminal and .env (in the repositorie that i want to extend the base project). and it doesn't worked.
My question is, there is a better example of how can i do that? i've been strugling with it a lot of hours... Thank you!

**Comment by ErosBurastero** [comment]
Momently, i've turned the repositorie in public state again, it extends the repositorie but doesn't read Vuetify and Assets folder, there is a way to fix it? Thank you

Here is the problems: Failed to resolve import "vuetify/lib/styles/main.sass" from "virtual:nuxt:C:\Users\ErosB\portfolio3\Nuxt3Starter\.nuxt\css.mjs". Does the file exist? 

Failed to resolve import "vuetify" from "..\..\AppData\Local\Temp\c12\github_ErosBurastero_Nuxt3Base\plugins\vuetify.ts". Does the file exist?    

note: In the base repositorie it works fine.

Thank you

**Comment by pi0** [comment]
> something like: layerAlias: '@layer1', to be used also in paths within the layer...

This feature is exactly the named layers in the roadmap 👍🏼

**Comment by Dominic-Marcelino** [comment]
> This feature is exactly the named layers in the roadmap 👍🏼

Is there a rough roadmap for the named layers? (weeks or months?)
Really looking to import data from custom folders

BTW: great work! The new layers/extends makes a lot of things much easier! 🚀

**Comment by nathanchase** [comment]
It seems like adding `extends` - even to point to a local repository - has a pretty significant impact on Nitro's **_server_** build time.

With:
```js
extends: [
    '../base'
  ],
```
Running `'pnpm dev' = Nitro built in 11284 ms

Without:
Running `'pnpm dev' = Nitro built in 411 ms

I created a barebones reproduction here: https://github.com/nathanchase/nuxt-layers-test

1. Git clone https://github.com/nathanchase/nuxt-layers-test.git
2. `pnpm i` inside `/app` and `/base`
3. `pnpm dev` inside `/app` - notice 11s server build time
4. Comment out `extends` block from `/app/nuxt.config.ts`
5. `pnpm dev` inside `/app` - notice <1s server build time

**Comment by pi0** [comment]
Thanks for reproduction @nathanchase I've moved it to https://github.com/nuxt/framework/issues/9947 to track better if it is a bug.

**Comment by hanzoz** [comment]
From what I have tested so far for private gitlab repo is

1. I need to set `export NODE_EXTRA_CA_CERTS=/path/to/cert.pem` in `.zshrc`
2. For example, my private repo is a self hosted gitlab then I need to modify the domain on line 87 & 88 in `/giget/shared/gitget.xxxx.mjs`
    url: `https://gitlab.domain.com/${parsed.repo}/tree/${parsed.ref}${parsed.subdir}`,
    tar: `https://gitlab.domain.com/${parsed.repo}/-/archive/${parsed.ref}.tar.gz`
 
_is there anything I can do for gitlab custom domain, for example to set GIGET_GITLAB_URL in .env so that it will pick up custom domain?_  

 3. set `GIGET_REGISTRY`, `GIGET_AUTH` in .env
 
So far the fetch in sendFetch function on line 44 in `/giget/shared/gitget.xxxx.mjs` is not working well, even though I provided my ca cert to .zshrc and.env config. It will stop downloading the main.tar.gz when access denied to the private repo. There will be an error showing "TAR_BAD_ARCHIVE: Unrecognized archive format".

any help on this?

**Comment by xibman** [comment]
Hello and first of all thank you for your great work.

After some tests on the layers which are bluffing I found an edge case I think.

Here is the link to reproduce it: https://stackblitz.com/edit/github-iabnrt

The base project contains the following nested routes

/nested (with an index content)
 /nested/test
/nested/test2


If you use this base layer and try to override the nested.vue page you lose the child routes from the base layer:

``` javascript
{
    name: 'nested',
    path: '/nested',
    file: '/home/projects/github-iabnrt/pages/nested.vue',
    children: [],
    meta: {},
    alias: [],
    redirect: undefined,
    component: [Function: component]
  }
```
Is there a way to handle route merge? 

Thanks for all again !

**Comment by OlaAlsaker** [comment]
Looks like extending nested pages is not handled very well.

**Example**
https://stackblitz.com/edit/github-iabnrt-9ruwhi

I have created a nested settings-page where the parent is `settings.vue`, and the child is `general.vue`. I have also created a "module" called `account` that I extend in the main project. The idea is that each module can provide it's own child-page in settings. So, if you look in `modules/account/pages/settings` you will see a `account.vue` file.

The problem occurs when you go to `localhost:3000/settings/account` - the parent page is not being used, it just shows the child-page (`account.vue`). However, if i go to `/settings/general`, everything works fine.

**TLDR;** Parent-pages to do not apply to child-pages that are extended.

Is this a bug, or is it just not implemented yet? 😄 

I guess a workaround in the meantime could be to just use the `<component :is="...">` to get a settings-component from the module.
